client/src/main/scala/com.gu.contentapi.client/ContentApiBackoff.scala

```diff
@@ -0,0 +1,101 @@
+package com.gu.contentapi.client
+
+import java.util.concurrent.TimeUnit
+
+import com.gu.contentapi.client.model.{ContentApiRecoverableException, HttpResponse}
+
+import scala.concurrent.{ExecutionContext, Future}
+import scala.concurrent.duration.Duration
+
+abstract class ContentApiBackoff extends Product with Serializable { self =>
+
+  def increment: ContentApiBackoff = self match {
+    // check max retries reached
+    case Exponential(_, n, max) if n == max => RetryFailed(max)
+    case Multiple(_, n, max, _) if n == max => RetryFailed(max)
+    case Constant(_, n, max) if n == max => RetryFailed(max)
+    // setup next delay cycle
+    case Exponential(d, n, max) =>
+      val delay = if (n == 0) Duration(d.toMillis, TimeUnit.MILLISECONDS) else Duration(Math.pow(2, n) * d.toMillis, TimeUnit.MILLISECONDS)
+      Exponential(delay, n + 1, max)
+    case Multiple(d, n, max, f) =>
+      val delay = if (n == 0) Duration(d.toMillis, TimeUnit.MILLISECONDS) else Duration(f * d.toMillis, TimeUnit.MILLISECONDS)
+      Multiple(delay, n + 1, max, f)
+    case Constant(d, n, max) =>
+      Constant(d, n + 1, max)
+    case x => x
+  }
+
+  def currentState: ContentApiBackoff = self
+
+  def execute(operation: => Future[HttpResponse])(implicit context: ExecutionContext): Future[HttpResponse] = {
+    self match {
+      case r: Retryable if(r.attempts == 0) => ContentApiBackoff.attempt(self.increment, operation)
```

Comment:
I don't think this is correct 🤔 Can you walk me through your reasoning? I may have missed something.

Reply:
`Retryable` is a superclass for all the backoff types except failure. Rather than have to inspect each possible type of backoff (`Exponential`, `Multiple` or `Constant`) we can just ask is this a `Retryable` that will be in one of two states, or is it a `RetryFailed`?